Add pwnlib.filesystem

heapcrash opened the pull request:
In lots of the code (particularly in the GDB module) we need to do a lot of juggling and `if ssh: foo` which is cumbersome.

This introduces a new module, `pwnlib.io`, which has a `File` and `NamedTemporaryFile` member, which behave like `file()` and `tempfile.NamedTemporaryFile()` except the also accept a `ssh=` argument which makes all of the `.read()` and `.write()` and `.chmod()` calls go over the SSH tube.

I dislike using names of existing functions, since it may make the code misleading, but also don't know what else to name them so that there's no confusion.

heapcrash committed:
[io] Introduce first draft of IO module

This is designed to make it easier to do file management when a piece of code
has to deal with creating a file that might need to be local, OR might need
to reside on some SSH server somewhere.

Specifically, the GDB code needs this for future enhancements

heapcrash committed:
[docs] Add documentation and doctests

heapcrash committed:
Update CHANGELOG

heapcrash commented:
@Arusekk I want your feedback on what to actually name these classes, I'm not super comfortable with `File` and `NamedTemporaryFile`

Arusekk commented:
How about moving around the functions `read` and `write` from `pwnlib.util.misc` to some, say, `pwnlib.filesystem`, and also doing `from os import chmod, mkdir, ...similar stuff...` (and maybe `from pwnlib.tubes.process import process`) there, and then doing `fs = ssh or pwnlib.filesystem`, and then doing `fs.chmod()`, `fs.mkdir()`, `fs.process()` and stuff? This would look convenient and abstract away the ssh vs local difference (we might even add some of it to adb, just as a best-noeffort, or maybe not).

heapcrash commented:
Sounds good!  Not sure I want to add the `.process()` bit though the rest sound good.

heapcrash commented:
@Arusekk what if we use this opportunity to add `pathlib` (Python3 only, but there is a Python2 backport (https://github.com/mcmtroffaes/pathlib2) to Pwntools and we could build on top of that?  e.g. we could have a pathlib.Path that will `repr()` as `ssh://foo@bar/path/to/file` and subclass `pathlib.Path` to have e.g. `pwnlib.filesystem.SSHPath` but in the GDB module we can just treat everything as a `Path` object and not worry about the underlying transport.

heapcrash committed:
Add pathlib2 requirement for Python2

heapcrash committed:
Use Python3 documentation

heapcrash committed:
Remove pwnlib.io module

heapcrash committed:
[filesystem] Add filesystem module, with Path and SSHPath

heapcrash committed:
Use @classmethod monkey patching instead of subclassing

This means that isinstance(Path, ssh_path_instance) returns True

heapcrash committed:
Make doctests work again

heapcrash committed:
Remove debugging code

heapcrash committed:
Rename context.ssh to context.ssh_session

This is needed because @LocalContext will pop off 'ssh' and some functions
take 'ssh' as an argument, and it would all break.

heapcrash committed:
[debug] Try to figure out why CI is failing

heapcrash committed:
[ssh] Fix failing doctest

heapcrash committed:
Fix README

heapcrash committed:
Fix README

heapcrash committed:
Fix failing doctest harder

heapcrash committed:
Remove unneeded internal file

heapcrash commented:
CI all passes, including the new tests.

Lint errors are mostly wrong.  Merging to dev so I can use this in the `gdb.attach` rewrite.

peace-maker commented on pwnlib/context/__init__.py:
```diff
@@ -1406,6 +1407,15 @@ def cyclic_size(self, size):
         return size
 
+    @_validator
+    def ssh_sesion(self, shell):
```
session*
